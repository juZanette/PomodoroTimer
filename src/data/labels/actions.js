import fs from '../../firebase/firebase'

export const setFullscreenDialog = (fullscreenDialog) => ({
  type: 'SET_FULLSCREEN_DIALOG',
  fullscreenDialog,
})

export const setDesktopDialog = (desktopDialog) => ({
  type: 'SET_DESKTOP_DIALOG',
  desktopDialog,
})

export const setDeleteAlert = (deleteAlert) => ({
  type: 'SET_DELETE_ALERT',
  deleteAlert,
})

export const setMenuOpened = (menuOpened) => ({
  type: 'SET_MENU_OPENED',
  menuOpened,
})

export const setFormValue = (formValue) => ({
  type: 'SET_FORM_VALUE',
  formValue,
})

export const setLabelEditting = (labelEditting) => ({
  type: 'SET_LABEL_EDITTING',
  labelEditting,
})

export const setLabelSelected = (labelSelected) => ({
  type: 'SET_LABEL_SELECTED',
  labelSelected,
})

export const addLabel = (label) => ({
  type: 'ADD_LABEL',
  label,
})

export const startAddLabel = (label) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    const ref = await fs.collection(`users/${uid}/labels`).add(label)

    dispatch(
      addLabel({
        id: ref.id,
        ...label,
      })
    )

    return ref
  }
}

export const editLabel = (id, updates) => ({
  type: 'EDIT_LABEL',
  id,
  updates,
})

export const startEditLabel = (id, updates) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    await fs.doc(`users/${uid}/labels/${id}`).update({
      ...updates,
    })

    dispatch(editLabel(id, updates))
  }
}

export const deleteLabel = (id) => ({
  type: 'DELETE_LABEL',
  id,
})

export const startDeleteLabel = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid
    await fs.doc(`users/${uid}/labels/${id}`).delete()

    dispatch(deleteLabel(id))
  }
}

export const setLabels = (labels) => ({
  type: 'SET_LABELS',
  labels,
})

export const startSetLabels = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    const labelsRef = await fs.collection(`users/${uid}/labels`).get()

    const labels = labelsRef.docs.map((label) => ({
      id: label.id,
      ...label.data(),
    }))

    dispatch(setLabels(labels))

    return labelsRef
  }
}
