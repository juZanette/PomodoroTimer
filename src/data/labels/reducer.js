const initialState = {
  dialogOpened: false,
  labelToEdit: null,
  formValue: {
    name: null,
    color: null,
  },
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_DIALOG_OPENED':
      return {
        ...state,
        dialogOpened: action.dialogOpened,
      }
    case 'SET_LABEL_TO_EDIT':
      return {
        ...state,
        labelToEdit: action.labelToEdit,
      }
    case 'SET_FORM_VALUE':
      return {
        ...state,
        formValue: action.formValue,
      }
    default:
      return state
  }
}
