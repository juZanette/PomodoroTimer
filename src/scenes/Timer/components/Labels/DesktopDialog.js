import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogTitle from '@material-ui/core/DialogTitle'
import { LabelForm } from './LabelForm'
import {
  setDesktopDialog,
  startAddLabel,
  startEditLabel,
  setFormValue,
  setLabelEditting,
} from '../../data/labels/actions'

export const DesktopDialog = () => {
  const { desktopDialog, formValue, labelEditting } = useSelector(
    (state) => state.labels
  )

  const dispatch = useDispatch()

  const handleClose = () => {
    dispatch(setDesktopDialog(false))
    dispatch(setLabelEditting(null))
    dispatch(setFormValue(null))
  }

  const handleConfirm = () => {
    if (labelEditting) {
      dispatch(startEditLabel(labelEditting.id, formValue))
    } else {
      dispatch(startAddLabel(formValue))
    }

    handleClose()
  }

  return (
    <div>
      <Dialog
        open={desktopDialog}
        onClose={handleClose}
        aria-labelledby="form-dialog-title"
        maxWidth="xs"
        fullWidth={true}
      >
        <DialogTitle id="form-dialog-title">Edit Label</DialogTitle>
        <DialogContent>
          <LabelForm />
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose} color="primary">
            Cancel
          </Button>
          <Button
            disabled={!formValue || !formValue.name || !formValue.color}
            onClick={handleConfirm}
            color="primary"
          >
            Confirm
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  )
}
