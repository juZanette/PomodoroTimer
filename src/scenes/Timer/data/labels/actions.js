export const setDialogOpened = (dialogOpened) => ({
  type: 'SET_DIALOG_OPENED',
  dialogOpened,
})

export const setAlertOpened = (alertOpened) => ({
  type: 'SET_ALERT_OPENED',
  alertOpened,
})

export const setLabelToEdit = (labelToEdit) => ({
  type: 'SET_LABEL_TO_EDIT',
  labelToEdit,
})

export const setFormValue = (formValue) => ({
  type: 'SET_FORM_VALUE',
  formValue,
})
