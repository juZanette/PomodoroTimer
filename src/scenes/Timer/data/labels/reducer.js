const initialState = {
  dialogOpened: false,
  formDialogOpened: false,
  deleteAlert: {
    opened: false,
    labelToDelete: null,
  },
  labelToEdit: null,
  formValue: null,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_DIALOG_OPENED':
      return {
        ...state,
        dialogOpened: action.dialogOpened,
      }
    case 'SET_FORM_DIALOG':
      return {
        ...state,
        formDialogOpened: action.formDialogOpened,
      }
    case 'SET_DELETE_ALERT':
      const { opened, labelToDelete } = action.deleteAlert

      return {
        ...state,
        deleteAlert: {
          opened,
          labelToDelete,
        },
      }
    case 'SET_LABEL_TO_EDIT':
      return {
        ...state,
        labelToEdit: action.labelToEdit,
      }
    case 'SET_FORM_VALUE':
      return {
        ...state,
        formValue: action.formValue,
      }
    default:
      return state
  }
}
