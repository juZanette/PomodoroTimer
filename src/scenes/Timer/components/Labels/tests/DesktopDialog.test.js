import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { createMount } from '@material-ui/core/test-utils'
import Button from '@material-ui/core/Button'
import labels from '../../../../../data/labels/tests/mock-data/labels'
import fs from '../../../../../firebase/firebase'
import { DesktopDialog } from '../DesktopDialog'

describe('<DesktopDialog />', () => {
  const mockStore = configureMockStore([thunk])

  const uid = 'asdf2222'
  let store
  let storeData
  let mount
  let wrapper

  beforeEach(async (done) => {
    const batch = fs.batch()

    const labelsRef = await fs.collection(`users/${uid}/labels`).get()

    labelsRef.docs.forEach(({ id }) => {
      batch.delete(fs.doc(`users/${uid}/labels/${id}`))
    })

    labels.forEach(({ id, name, color }) => {
      const ref = fs.collection(`users/${uid}/labels`).doc(id)
      batch.set(ref, { name, color })
    })

    await batch.commit()
    done()

    mount = createMount()

    wrapper = (labelEditting = null, formValue = null) => {
      storeData = {
        auth: { uid },
        labels: {
          desktopDialog: true,
          labelEditting,
          formValue,
          data: labels,
        },
      }

      store = mockStore(storeData)
      store.dispatch = jest.fn()

      return mount(
        <Provider store={store}>
          <DesktopDialog />
        </Provider>
      )
    }
  })

  afterEach(() => {
    mount.cleanUp()
  })

  test('should render <DesktopDialog /> correctly', () => {
    const wrapperRendered = wrapper(null, {
      name: 'Play Piano',
      color: '#123123',
    })

    expect(wrapperRendered).toMatchSnapshot()
  })

  test('Confirm button should be disable when incomplete data is provided', () => {
    const wrapperRendered = wrapper(null, {
      name: 'Play Piano',
    })

    expect(wrapperRendered.find(Button).at(1).prop('disabled')).toBe(true)
  })

  test('Confirm button should be enabled when complete data is provided and should call dispatch 4 times on confirm button click', () => {
    const wrapperRendered = wrapper(null, {
      name: 'Play Piano',
      color: '#123123',
    })

    expect(wrapperRendered.find(Button).at(1).prop('disabled')).toBe(false)
    wrapperRendered.find(Button).at(1).simulate('click')
    expect(store.dispatch).toHaveBeenCalledTimes(4)
  })
})
