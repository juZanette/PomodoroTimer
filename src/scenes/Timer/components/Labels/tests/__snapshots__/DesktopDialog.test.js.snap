// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DesktopDialog /> should render <DesktopDialog /> correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Dialog))
    aria-labelledby="form-dialog-title"
    fullWidth={true}
    maxWidth="xs"
    onClose={[Function]}
    open={true}
  >
    <WithStyles(ForwardRef(DialogTitle))
      id="form-dialog-title"
    >
      Edit Label
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))>
      <LabelForm />
    </WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogActions))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        disabled={false}
        onClick={[Function]}
      >
        Confirm
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(DialogActions))>
  </WithStyles(ForwardRef(Dialog))>
</div>
`;
